x-pack/plugins/apm/server/lib/helpers/transaction_coldstart_rate.ts

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { FAAS_COLDSTART } from '../../../common/elasticsearch_fieldnames';
+import {
+  AggregationOptionsByType,
+  AggregationResultOf,
+} from '../../../../../../src/core/types/elasticsearch';
+
+export const getColdstartAggregation = () => ({
+  terms: {
+    field: FAAS_COLDSTART,
+  },
+});
+
+type ColdstartAggregation = ReturnType<typeof getColdstartAggregation>;
+
+export const getTimeseriesAggregation = (
+  start: number,
+  end: number,
+  intervalString: string
+) => ({
+  date_histogram: {
+    field: '@timestamp',
+    fixed_interval: intervalString,
+    min_doc_count: 0,
+    extended_bounds: { min: start, max: end },
+  },
+  aggs: { coldstartStates: getColdstartAggregation() },
+});
+
+export function calculateTransactionColdstartRate(
+  coldstartStatesResponse: AggregationResultOf<ColdstartAggregation, {}>
+) {
+  const coldstartStates = Object.fromEntries(
+    coldstartStatesResponse.buckets.map(({ key, doc_count: count }) => [
+      key === 1 ? 'true' : 'false',
+      count,
+    ])
+  );
+
+  const coldstarts = coldstartStates.true ?? 0;
+  const warmstarts = coldstartStates.false ?? 0;
+
+  return coldstarts / (coldstarts + warmstarts);
```

Comment:
There's probably a more straight-forward way to express this. What about using something like lodash's `partition`?

```ts
const [ warmStarts, coldStarts ] = partition(coldstartStatesResponse, ( bucket ) => bucket.key === 1);

return coldStarts.length / (coldStarts.length - warmStarts.length);
```